openshift-ci: adapt scripts to test like OpenShift Sandboxed Containers

wainersm opened the pull request:
This is the needed changes on Kata Containers side to allow:
1) build with same configuration as OpenShift Sandboxed Containers
2) configure the runtime to use the host kernel
3) configure the runtime to use the system QEMU

Fixes #4163

wainersm commented:
@fidencio if you have some free cycles then could you help with your review? thx!

fidencio commented:
> @fidencio if you have some free cycles then could you help with your review? thx!

In general I like the idea of testing "middlestream" as much as possible, but I'd like to be sure we're both on the same page here.

Enabling the extension will install kata-containers & qemu on the node.
Installing any RPM package on RHCOS will require a system reboot so the package can be consumed.
A system reboot will automatically generate the kernel & initrd you'd like to use.

With this in mind, I'd say we could rely more on the work done on RHCOS for this, by:
* Installing the kata-containers RPM from the extensions (remember, it'll trigger a reboot).
* Once the reboot happened, build kata-containers from source and "install" binaries on the very same location used by the original ones (including the agent), and remember, you can always remount the `ro` filesystem as `rw` for testing purposes.
* Delete the content from `/var/cache/kata-containers` and retrigger the `kata-osbuilder-generate.service`

By doing this we'd be able to have the initrd automagically generated using the new agent.

One thing that's making me scratch my head is whether you're fine with the reboot or not.  If a reboot can't happen, we'll have to think about a workaround here as:
* Installing new packages would require a reboot (that can be worked around).
* Adding a CRI-O drop-in *may* require a reboot as well (@saschagrunert, @haircommander, is this the case? if not we can simply restart reload CRI-O).

Jakob-Naucke commented on .ci/openshift-ci/build_install.sh:
```diff
@@ -19,12 +19,68 @@ source /etc/os-release
+sandboxed_containers_image_configs() {
+	# It needs to overwrite some variables of
+	# kata-containers/tools/osbuilder/Makefile.
+	export DRACUT_KVERSION="4.18.0-305.19.1.el8_4.x86_64"
```
Isn't this more fitting for `versions.yaml`?

wainersm replied:
Good idea @Jakob-Naucke ! I've been trying to not spill OpenShift CI specific code out of the `.ci/openshift-ci` folder but I think a new entry on `versions.yaml` doesn't hurt :)

wainersm replied:
Done.

Jakob-Naucke commented on .ci/openshift-ci/cluster/install_kata.sh:
```diff
@@ -66,6 +75,52 @@ wait_for_reboot() {
+		"$degraded_count" == 0 ]]; do
+		# Let's check it hit the timeout (or not).
+		local now=$(date +%s)
+		if [ $(($timer_start + $delta)) -lt $now ]; then
```
I guess you wouldn't really have to run this addition every time but it doesn't matter a ton

wainersm replied:
True. I will change it.

wainersm replied:
Done.

Jakob-Naucke commented on .ci/openshift-ci/cluster/deployments/daemonset_kata-installer.yaml.in:
```diff
@@ -35,6 +35,18 @@ spec:
+                      name: ci.kata.sandboxedcontainers
+                      key: qemu_path
+                      optional: true
+                - name: USE_HOST_KERNEL
```
Wouldn't this become `yes` (and if so, is that correct)?

wainersm replied:
Hi @Jakob-Naucke , I am not sure I understood your question. The value of `USE_HOST_KERNEL` is read from a optional `configMap`; if the `configMap` is not deployed then `USE_HOST_KERNEL` will be exported to the container as an empty variable.

Jakob-Naucke replied:
ah, right. Nevermind.

Jakob-Naucke reviewed (commented):
Thank you @wainersm -- looks pretty clean overall. I have a small number of questions...

jodh-intel commented on .ci/openshift-ci/build_install.sh:
```diff
@@ -19,12 +19,68 @@ source /etc/os-release
+	cp -r "${kata_repo_dir}/tools/osbuilder/dracut/dracut.conf.d"/* \
+		"${DRACUT_CONF_DIR}"
+	cat <<-EOF >> "${DRACUT_CONF_DIR}/10-drivers.conf"
+	drivers+="vmw_vsock_virtio_transport virtio_net virtiofs "
+	drivers+="virtio_blk virtio_scsi virtio_console "
+	drivers+="vfio-pci vfio vfio_iommu_type1 irqbypass vfio_virqfd "
```
It would spread the code over extra lines, but it might be better to add one driver per line. That makes diffs easier for future changes and also emphasises that crucial space at the end of each of these lines. We can also then sort the list alphabetically ;)

wainersm replied:
@jodh-intel I agree. I will change that.

wainersm replied:
Done.

haircommander commented:
>     * Adding a CRI-O drop-in _may_ require a reboot as well (@saschagrunert, @haircommander, is this the case? if not we can simply restart reload CRI-O).

A node reboot is excessive for CRI-O. At most extreme case all we need is to restart the crio process (`systemctl restart crio`), and possibly a `crictl rmp -fa` if we want to clear the container state

wainersm commented on .ci/openshift-ci/cluster/deployments/daemonset_kata-installer.yaml.in:
```diff
@@ -35,6 +35,12 @@ spec:
+                - name: QEMU_PATH
+                  valueFrom:
+                    configMapKeyRef:
+                      name: ci.kata.sandboxedcontainers
```
fyi: just realized `name` should be `ci.kata.installer`. I will fix it.

wainersm replied:
Fixed. I am actually now using two config maps: `ci.kata.installer.qemu`  and `ci.kata.installer.kernel` as I realized updating a configMap is a non-trivial task.

snir911 commented on kata-webhook/deploy/webhook.yaml:
```diff
@@ -24,7 +24,11 @@ spec:
+                configMapKeyRef:
+                  name: kata-webhook
+                  key: runtime_class
+                  optional: true
```
Is it still uses "kata" as default? can it be done?

wainersm replied:
Good point @snir911. I was looking at pod's `env` specification (https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#environment-variables-1), it is not possible to set a default value. `configMapKeyRef` also doesn't allow a default value. Note that in https://github.com/kata-containers/tests/blob/main/kata-webhook/main.go#L77 it is going to use `kata` if `RUNTIME_CLASS` is not export.

Anyway, I need to test it works fine if the configMap isn't deployed.

wainersm replied:
@snir911 I tested it locally; the change proposed on kata-webhook/deploy/webhook.yaml will not break the default workflow (i.e. when the configMap isn't deployed)

fidencio commented:
> > ```
> > * Adding a CRI-O drop-in _may_ require a reboot as well (@saschagrunert, @haircommander, is this the case? if not we can simply restart reload CRI-O).
> > ```
> 
> A node reboot is excessive for CRI-O. At most extreme case all we need is to restart the crio process (`systemctl restart crio`), and possibly a `crictl rmp -fa` if we want to clear the container state

Yeah, but I think that depending on how it's done (via the MCO, for instance) the MCO would trigger a reboot regardless.

wainersm commented:
/test-fedora

wainersm commented:
Hi @fidencio , thanks for the comments. Here goes my reply.

> > @fidencio if you have some free cycles then could you help with your review? thx!
> 
> In general I like the idea of testing "middlestream" as much as possible, but I'd like to be sure we're both on the same page here.
> 

I'm glad to hear that :)

> Enabling the extension will install kata-containers & qemu on the node. Installing any RPM package on RHCOS will require a system reboot so the package can be consumed. A system reboot will automatically generate the kernel & initrd you'd like to use.
> 
> With this in mind, I'd say we could rely more on the work done on RHCOS for this, by:
> 
>     * Installing the kata-containers RPM from the extensions (remember, it'll trigger a reboot).
> 
>     * Once the reboot happened, build kata-containers from source and "install" binaries on the very same location used by the original ones (including the agent), and remember, you can always remount the `ro` filesystem as `rw` for testing purposes.
> 
>     * Delete the content from `/var/cache/kata-containers` and retrigger the `kata-osbuilder-generate.service`
> 
> 
> By doing this we'd be able to have the initrd automagically generated using the new agent.
> 

We talked in private about this PR but allow me to explain to the public some decisions I took.

I'm installing the RHCOS extension so that QEMU gets installed in the nodes;  the kata-containers rpm is installed as a "side effect". I don't know if the nodes are reboot or not, and it doesn't matter to me because I am not using the service that rebuilds the initrd. In order to use that service I would had to overwrite the kata-containers rpm files, which I am not doing because IMHO it is error-prone. Instead I am installing my built kata containers in `/opt` (i.e. on the node there will be two kata installations coexisting).

> One thing that's making me scratch my head is whether you're fine with the reboot or not. If a reboot can't happen, we'll have to think about a workaround here as:
> 
>     * Installing new packages would require a reboot (that can be worked around).
> 
>     * Adding a CRI-O drop-in _may_ require a reboot as well (@saschagrunert, @haircommander, is this the case? if not we can simply restart reload CRI-O).

Currently I add a CRI-O drop-in (via MachineConfig) to configure the runtime class and this has triggered a reboot.

wainersm commented:
@snir911 @Jakob-Naucke @jodh-intel I addressed your comments. Can I have your approval or another round of reviews?

@fidencio are you okay if I merge that implementation which doesn't use the downstream initrd rebuild service? (maybe I can implement that in the future, if worth it).

jodh-intel commented on .ci/openshift-ci/cluster/install_kata.sh:
```diff
@@ -66,6 +75,52 @@ wait_for_reboot() {
+		# Let's check it hit the timeout (or not).
+		local now=$(date +%s)
+		if [ $deadline -lt $now ]; then
+			echo "Timeout: not all workers updated"
```
Nit: Output to stderr?

wainersm replied:
done.

jodh-intel commented on .ci/openshift-ci/cluster/install_kata.sh:
```diff
@@ -66,6 +75,52 @@ wait_for_reboot() {
 
+wait_mcp_update() {
+	local delta="${1:-900}"
+	local sleep_time=60
```
That's a pretty long sleep!

wainersm replied:
@jodh-intel ,

Have you ever watched TV series like [Pawn Stars](https://www.imdb.com/title/tt1492088/)? At some point they "meet in the middle" and shake hands. So 30s of sleep time works for you? :)

jodh-intel commented on .ci/openshift-ci/images/entrypoint.sh:
```diff
@@ -32,11 +37,37 @@ function terminate()
 
+# Set the QEMU path on configuration.toml to $QEMU_PATH, also update
+# virtiofsd's path.
+function update_qemu_path()
```
Nit: The `function` prefix is consistent with this file, but looks odd on this PR since other functions in other files don't use it. It's entirely redundant afaik so maybe drop it?

wainersm replied:
done

jodh-intel replied:
On that and related shell topics, any thoughts on https://github.com/kata-containers/kata-containers/pull/3100#issuecomment-976405250 ?

jodh-intel reviewed (approved):
Thanks @wainersm.

lgtm

wainersm commented:
/test

wainersm commented:
@snir911 @Jakob-Naucke one more ACK please? :)

wainersm commented:
/test-firecracker

snir911 commented on .ci/openshift-ci/build_install.sh:
```diff
@@ -60,6 +60,37 @@ sandboxed_containers_build_configs() {
+	drivers+="virtio_scsi "
+	drivers+="vmw_vsock_virtio_transport "
+	dracutmodules+=" bash rescue "
+	EOF
```
I wondered if we can pull this file from somewhere, can be done later though..

wainersm replied:
Good idea. I will send a follow up pull request

snir911 reviewed (approved):
LGTM thanks!

wainersm commented:
@Jakob-Naucke @snir911 @jodh-intel thanks for the reviews!
